crates/revm/src/executor.rs

```diff
@@ -431,35 +431,48 @@ pub fn commit_state_changes<DB>(
                 Entry::Vacant(entry) => {
                     let entry = entry.insert(Default::default());
                     entry.info = account.info.clone();
-
-                    let account = to_reth_acc(&entry.info);
-                    if !(has_state_clear_eip && account.is_empty()) {
-                        post_state.create_account(block_number, address, account);
+                    entry.account_state = AccountState::NotExisting; // we will promote account state down the road
+                    let new_account = to_reth_acc(&entry.info);
+
+                    #[allow(clippy::nonminimal_bool)]
+                    // If account was touched before state clear EIP, create it.
+                    if !has_state_clear_eip ||
+                        // If account was touched after state clear EIP, create it only if it is not empty.
+                        (has_state_clear_eip && !new_account.is_empty())
+                    {
+                        post_state.create_account(block_number, address, new_account);
                     }
+
                     entry
                 }
                 Entry::Occupied(entry) => {
                     let entry = entry.into_mut();
 
-                    if entry.info.is_empty() {
-                        let account = to_reth_acc(&account.info);
-                        if !(has_state_clear_eip && account.is_empty()) {
-                            post_state.create_account(block_number, address, account);
-                        }
-                    } else if entry.info != account.info {
+                    let old_account = to_reth_acc(&entry.info);
+                    let new_account = to_reth_acc(&account.info);
+
+                    let account_non_existent =
+                        matches!(entry.account_state, AccountState::NotExisting);
+
+                    // Before state clear EIP, create account if it doesn't exist
+                    if (!has_state_clear_eip && account_non_existent)
+                        // After state clear EIP, create account only if it is not empty
+                        || (has_state_clear_eip && entry.info.is_empty() && !new_account.is_empty())
+                    {
+                        post_state.create_account(block_number, address, new_account);
```

Comment:
Ok, if the account does not exist or account is empty before state clear, we are creating new account, makes sense.